objects/src/accounts/code.rs

```diff
@@ -0,0 +1,79 @@
+use super::{AccountError, Digest};
+use assembly::{parse_module, ModuleAst};
+use crypto::merkle::MerkleTree;
+use miden_core::Program; // TODO: we should be able to import it from the assembly crate
+
+// ACCOUNT CODE
+// ================================================================================================
+
+/// Describes public interface of an account.
+///
+/// Account's public interface consists of a set of Miden VM programs each defined its own MAST.
+/// Root of each MAST commits to the underlying program. We commit to the entire account interface
+/// by building a simple Merkle tree out of all procedure MAST roots.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AccountCode {
+    method_tree: MerkleTree,
+    module: ModuleAst,
+    // methods: Vec<Program>, commented out because Program doesn't currently implement Eq and
+    // PartialEq. Also, there might be a better way of describing a set of programs as they
+    // might share a lot of common code blocks. In a way, we want something like a Program
+    // struct but with many entry points.
+}
+
+impl AccountCode {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Creates and returns a new definition of an account's interface compiled from the specified
+    /// source code.
+    pub fn new(source: &str) -> Result<Self, AccountError> {
+        let _module_ast = parse_module(source)?;
+
+        // TODO: compile module AST into a set of program MASTs. To do this we need to expose
+        // a new method on the assembler to compile a module rather than a program (something
+        // similar to Assembler::compile_module() but without internal parameters).
+        //
+        // Open question: how to initialize the assembler? Specifically, which libraries to
+        // initialize it with. stdlib and midenlib are the two libraries we need for sure - but
+        // how to handle accounts which rely on some user-defined libraries? i.e., should there
+        // be a way to specify an "on-chain" library somehow?
```

Comment:
being able to specify an "on-chain" library seems like it would give some composability benefits

Reply:
One idea is to use the rollup as a data layer / key-value store:

- For point-to-point communication, where the parties do not have a out-of-band communication channel, Alice could encrypt the data to be delivery to Bob, and publish the encrypted data to the rollup.
  - I thought about using this in the context of [user-to-user](https://github.com/0xPolygonMiden/miden-base/discussions/11) flow.
  - The public key itself could also be register on the rollup.
- This could also be a way of revealing data, including code.
  - I thought this on the context of [hidden code](https://github.com/0xPolygonMiden/miden-base/discussions/7#discussioncomment-4489315)
  - One benefit of this is that it allows code reuse, for instance, most likely there will be a standard for fungible assets, and a repository with sample code to build from. I assume a few of these methods will be the same, meaning they can share the same piece of code from the repository. (Assuming a somewhat stable output from the existing compilers)

So basically we would publish the MAST to the rollup as data, the key would be the MAST root, the contents the MAST itself. Anyone would be allowed to publish such data, and any note would be allowed to use the registered code.

There are tons of open questions, like how to decided how long the data is kept, how to whitelist code, how to derive the key from the content in a way that the above things are support. But this is just an idea.

Another idea is to have an idea of a versioned prelude defined as part of the chain protocol, each note would define which prelude to use. That way we can rollout new functionality in a backwards compatible way.